docs/deprecation-process.md

```diff
@@ -0,0 +1,309 @@
+  `DEPRECATED` with the new release date,
+* In the `CHANGELOG` and the release notes, add references
+  to the new deprecated items for this release,
+  pointing to the `DEPRECATED` document.
+
+### Wait
+
+Do not implement the removal right away.
+
+First, if a removal is implemented and merged to main just after a new release,
+it exposes itself to a full revert, should another release be necessary to fix
+any regressions just caused recently.
+
+Second, some people will only notice the deprecation when discovering it
+in the release, no matter how many previous announcements were done.
+Allow some time for people to raise issues or concerns,
+especially if there are special usage patterns that were not anticipated.
+
+Once things are stable, proceed with the removal.
+Depending on the change, it can be implemented:
+
+* in the next minor release,
+* after a few minor releases,
+* in the next major release,
+* after a few major releases
+
+following the deprecation.
+
+The more disruptive the change is,
+the more time users will need to adjust,
+so trivial deprecations can be removed right away,
+while complex ones can take much longer.
+
+In any case, never implement the deprecation and the removal in the same
+release: wait.
+
+### Prepare removal
+
+Implement the removal issue with a PR.
+
+Remove all the deprecated code.
+
+Remove all the tooling (compiling options) related to the deprecated code,
+if any.
+
+Remove all the relevant entries in `DEPRECATED`.
+
+Add a `CHANGELOG` entry for the removal.
+
+Get the removal PR reviewed and merged to the main branch.
+
+### Release removal
+
+When making a new opentelemetry-cpp release,
+in addition to the regular release tasks:
+
+* In the `CHANGELOG` and the release notes, add references
+  to the new removed items for this release.
+
+## Technical guidelines
+
+### CMake options
+
+Assume the option `WITH_FOO` needs to be deprecated.
+
+Code using `WITH_FOO=OFF` or `WITH_FOO=ON` must build as before,
+yet users should be notified if they use `WITH_FOO` in their build.
+
+Assume a build option `WITH_NO_DEPRECATION` already exists.
+
+In a normal build, used in production, code is compiled with
+`WITH_NO_DEPRECATION=OFF`.
+
+In a verification build, code is compiled with `WITH_NO_DEPRECATION=ON`.
+
+Implement the following logic in CMake:
+
+```cmake
+  option(WITH_FOO "DEPRECATED - With the foo feature" OFF)
+
+  if(WITH_FOO)
+    if(WITH_NO_DEPRECATION)
+      message(FATAL_ERROR "WITH_FOO is deprecated")
+    else()
+      message(WARNING "WITH_FOO is deprecated")
+    endif()
+  endif
+```
+
+The verification build is not meant to be used in production.
+It is meant to find all references of deprecated items.
+
+If the verification build succeeds, the user code is guaranteed to be clean,
+and will not be impacted by the removal to come.
+
+When implementing such logic, document it in the mitigation section,
+in file `DEPRECATED`.
```

Comment:
Do you think the `DEPRECATED.md` file is necessary, or we can continue to use `CHANGELOG.md` file to document both DEPRECATION and REMOVAL.
So for any new release, if the release includes any PR implementing deprecation, we can use [DEPRECATION] tag in the CHANGELOG to describe this. Most of the information as of now planned to be added in DEPRECATION.md (i.e., Motivation and Scope) can also be described in the PR/Issue whose reference/link would be there in CHANGELOG.md.
And once we release includes any PR removing the deprecated code, we can use [WARNING] or [REMOVAL] tag in the CHANGELOG to describe it.

Reply:
> Do you think the `DEPRECATED.md` file is necessary, or we can continue to use `CHANGELOG.md` file to document both DEPRECATION and REMOVAL. So for any new release, if the release includes any PR implementing deprecation, we can use [DEPRECATION] tag in the CHANGELOG to describe this. Most of the information as of now planned to be added in DEPRECATION.md (i.e., Motivation and Scope) can also be described in the PR/Issue whose reference/link would be there in CHANGELOG.md. And once we release includes any PR removing the deprecated code, we can use [WARNING] or [REMOVAL] tag in the CHANGELOG to describe it.

Hi @lalitb

We can and should keep the `CHANGELOG.md` file to describe what changed in a given release, with entries for **_new_** `DEPRECATION` and **_new_** `REMOVAL`. These entries should point to the details located in file `DEPRECATED.md`.

I think maintaining `DEPRECATED.md` is critical:

- anyone using opentelemetry version X for the first time should not have to inspect all the previous changelogs (back to when ?) to figure out what is and what is not deprecated in the release they use.
- `DEPRECATED.md` will list **_all_** deprecations that still apply.

Overall, deprecating something is a matter is mitigating frictions between maintainers (who want/need to change things) and users (who need bug fixes but otherwise want no changes).

By providing just one document, easy to read, and organized by clear sections (I use platform XYZ, any deprecation listed under the platform section ? No ? great, all set then.), I think it will be easier:

- for users to not miss important announcements, with one single place to check,
- for maintainers to have a clear picture of the migration path.

Reply:
I'd vote for having `DEPRECATED.md`.